doc/design/optimizer.md

```diff
@@ -0,0 +1,85 @@
+## Optimizer Design
+In deeplearning system, `Optimizer` is used to optimize(minimize) loss thow updating a list of parameters. 
+
+### A typical training process:
+
+1. run forward to calculate activation using data and parameter.
```

Comment:
I do not think this typical training process fits our current design.

Currently, we put every operator into one `ProgramDesc`.  There are not three running stages explicitly.

Reply:
This is a general abstract training process, no matter how complex the training process is, they are all composed of these stages. 

In Our design, we also have functions like backward and optimize to put related operators into ProgramDesc. Here we just put the interface into `Optimizer` as high level API.